stage/stage_updc.man
.\" $Id: stage_updc.man,v 1.1 2000/05/15 13:00:42 jdurand Exp $
.\"
.\" @(#)$RCSfile: stage_updc.man,v $ $Revision: 1.1 $ $Date: 2000/05/15 13:00:42 $ CERN IT-PDP/DM Jean-Damien Durand
.\" Copyright (C) 2000 by CERN/IT/PDP/DM
.\" All rights reserved
.\"
.TH STAGE_UPDC "3" "$Date: 2000/05/15 13:00:42 $" "CASTOR" "Stage Library Functions"
.SH NAME
stage_updc \- update the stager catalog
.SH SYNOPSIS
.BI "#include <" stage_api.h ">"
.sp
.BI "int stage_updc_tppos(char *" stageid ", int " status ", int " blksize ", char *" drive ", char *" fid ", int " fseq ", int " lrecl ", char *" recfm ", char *" path ");"
.sp
.BI "int stage_updc_filcp(char *" stageid ", int " status ", char *" ifce ", u_signed64 " size ", int " waiting_time ", int " transfer_time ", int " blksize ", char *" drive ", char *" fid ", int " fseq ", int " lrecl ", char *" recfm ", char *" path ");"

.SH DESCRIPTION
The \fBstage_updc_filcp\fP and \fBstage_updc_tppos\fP API methods calls for an update of a entry in the
.I stager
catalog.
.P
The \fBstage_updc_tppos\fP method is typically used to indicate to the stager that there has been a successful tape positionning, whereas \fBstage_updc_filcp\fB indicates to the stager that the current subrequest corresponding to global request
.BI stageid
is over.

.SH PARAMETERS
.TP 2.0i
.BI "char *" stageid
Gives the global stageid. This parameters is in the form: reqid.key@stagehost
.TP
.BI "int status"
Exit code for the file stage subrequest. It is used in \fBstage_updc_tppos\fP only if its value is \fBETFSQ\fP.
.TP
.BI "char *" ifce
Identifies the network interface used for the transfer between the tape server and the disk server.
.TP
.BI "u_signed64 " size
Gives the actual filesize in bytes.
.TP
.BI "int " waiting_time
Time spent before starting to transfer data between tape and disk server.
.TP
.BI "int " transfer_time
Time spent transferring data between tape and disk server.
.TP
.BI "int " blksize
Specifies the maximum block size.
.TP
.BI "char *" drive
Identifies the drive where the tape is mounted.
.TP
.BI "char *" fid
Specifies the fileid in HDR1.
.TP
.BI "int " fseq
Specifies the file sequence number of the tape file.
.TP
.BI "int " lrecl
Specifies the logical record length for fixed length record format files.
.TP
.BI "char *" recfm
Specifies the record format of the data on the tape. It may be
.BR U ,
.B F
or
.BR FB .
.TP
.BI "char *" path
Buffer where to receives enventual full path filename. Its recommended size is
.B CA_MAXPATHLEN+1.

.SH RETURN VALUE
0 on success, -1 on failure.

.SH ERRORS
If failure, the serrno variable might contain one of the following error codes:
.TP 1.9i
.B SENOMAPFND
Can't open mapping database
.TP
.B EFAULT
Bad address (invalid argument)
.TP
.B EINVAL
Invalid argument (bad stageid)


CHANGELOG.md
## 5.11.23.0-1 (2026-09-25)

### New Features

- [frontend] Add user configuration improvements for cta-admin-grpc (cta/CTA!1635)

## 5.11.22.0-1 (2026-09-04)

### New Features

- [frontend] Implement JWT blacklisting on the CTA Frontend (cta/CTA!1548)

- [runtime] Add additional validation logic to application config (cta/CTA!1567)


### Bug Fixes

- [scheduler] Fix issues after pg1 release rest: repack admin commands, archive...

- [Tools] Fix 'priority' column in 'cta-admin dr ls' command (cta/CTA!1546)

- [runtime] Make signals, telemetry and health handling more robust (cta/CTA!1554)

- [runtime] Add bounds checking and better error diagnostics to TOML loading (cta/CTA!1559)

## 5.11.21.0-1 (2026-08-25)

### New Features

- [scheduler] Add scheduler workflow metrics to OpenTelementry (cta/CTA!1387)


### Changes

- [Tools] Move cta-admin af ch command to tf ch (cta/CTA!1525)
- [Tools] Replace cta-objectstore-tools.conf with cta-scheduler.conf (cta/CTA!1540)
- [maintd] Change scheduler.objectstore_backend_path to scheduler.config_file (cta/CTA!1518)


### Bug Fixes

- [frontend] Fix logic bug in frontend authentication and SEGFAULT in cta-admin (cta/CTA!1558)
- [Tools] Fix cta-verify-file (cta/CTA!1545)
- [Tools] Add `--storage-class` option to `cta-send-event` (cta/CTA!1542)


### Performance Improvements

- [taped] Skip block reposition request if drive is already positioned correctly (cta/CTA!1527)

## 5.11.20.0-1

### New Features

- [Tools] Add optional storage class statistics, per tape, to cta-admin tool (cta/CTA!1410)
- [Misc] Add event_name to important CTA log messages (cta/CTA!1381)

### Changes

- [Tools] Change short option for cta-admin af ch --idfile from -I to -F (cta/CTA!1453)
- [taped] Taped no longer sets its own user, group and SYS_RAWIO capability (cta/CTA!1426)
- [Misc] Improve .service and .logrotate configuration (cta/CTA!1524)
- [Misc] Modernise RPM spec metadata and file handling (cta/CTA!1507)
- [Misc] Move systemd units to `/usr/lib/systemd/system` (cta/CTA!1497)
- [Misc] Change systemd to weak RPM dependency of CTA (cta/CTA!1509)
- [Misc] Change logrotate to weak RPM dependency of CTA (cta/CTA!1505)
- [Misc] Remove RPM changelog; package CHANGELOG.md as doc (cta/CTA!1508)
- [Misc] Remove noreplace from example configuration files (cta/CTA!1506)
- [taped] Drive process crashing produces CRIT message instead of INFO (cta/CTA!1499)
- [taped] cta-eosdf no longer sources sysconfig (cta/CTA!1455)

### Bug Fixes

- [frontend] Forbid reclaiming tapes with recently deleted files (cta/CTA!1418)
- [frontend] PREPARE gRPC request shouldn't fail without a storage class (cta/CTA!1471)
- [taped] Fix crash due to accidental pointer arithmetic in exception msg (cta/CTA!1502)
- [taped] Fix opentelemetry shutdown race condition with objectstore (cta/CTA!1500)
- [scheduler]: Fix archive unmount decision logic for postgres scheduler (cta/CTA!1450)
- [frontend] Fix frontend catalogue cache for mount rules (cta/CTA!1490)
- [CI] Ensure CTA images trust CERN Root CA (cta/CTA!1535)
- [Misc] Ensure CMake correctly installs executables as PROGRAMS (cta/CTA!1523)

### Security

- [CI] Generate SBOM for each Docker image (cta/CTA!1532)

### Other

- [CI] Add system tests to cover staging tokens on EOS-CTA (cta/CTA!1493)

## 5.11.19.0-1

### Changes

- [frontend] Split Admin/WFE frontends, add mTLS (cta/CTA!1367)
- [catalogue] Forbid adding or changing encryption key of non-empty tapes (cta/CTA!1414)

### Deprecations

- [frontend] Deprecate storage class field in EOS-CTA xattrs (cta/CTA!1424)

### Other

- [CI] Add Enstore and EnstoreLarge tape tests (cta/CTA!1412)

## 5.11.18.0-2

### Bug Fixes

- [Misc] Update required runtime version of opentelemetry-cpp to v1.26.0 (cta/CTA!1370)

## 5.11.18.0-1

### New Features

- [maintd] Add catalogue and telemetry conf files to runtime directory (cta/CTA!1303)
- [Misc] Package basic JSON logging schema (cta/CTA!1304)
- [Misc] Add source location and CTA version to log lines (cta/CTA!1348)

### Changes

- [frontend] Allow admin commands or physics workflows to be disabled on CTA frontend (cta/CTA!1336)
- [Misc] Replace reflect-cpp library with custom implementation (cta/CTA!1291)
- [Misc] Change Catalogue backend install behaviour (cta/CTA!1295)

### Bug Fixes

- [catalogue] Improve CTA resilience to catalogue DB outages during data fetching (cta/CTA!1334)
- [taped] Improve readExactBlock error message contents (cta/CTA!1311)
- [taped] Enabling reading of tapes with misformed OSM tape format (cta/CTA!1343)
- [frontend] Prevent 'cta-admin dr rm' command from crashing the cta-frontend (cta/CTA!1305)
- [frontend] Remove race conditions in NegotiationRequestHandler (cta/CTA!1310)
- [frontend] Fix broken cta-admin fr ls command (cta/CTA!1340)
- [Misc] Fix compile issues when building CTA in Release mode (cta/CTA!1207)

### Performance Improvements

- [CI] Use python xrootd persistent clients for archive instead of xrdcp --parallel (cta/CTA!1274)

### Other

- [Misc] Update opentelemetry-cpp to v1.26.0 (cta/CTA!1321)


## 5.11.17.0-1

### New Features

- [maintd] Add CLI option to use runtime directory for app metadata (cta/CTA!1266)

### Changes

- [maintd] Update maintd to use TOML configuration (cta/CTA!1246)
- [maintd] Use declarative config to initialise OpenTelemetry SDK (cta/CTA!1248)
- [maintd] Add use_unix_domain_socket configuration entry for health server (cta/CTA!1282)
- [maintd] Remove daemon_user and daemon_group settings (cta/CTA!1252)

### Bug Fixes

- [frontend] Fix crash in 'cta-admin dr ls' command when using multiple...
- [frontend] Add error message when trying to change media type of non-empty tape (cta/CTA!1253)

### Removals

- [frontend] Stop using TAPE_POOL.SUPPLY column (cta/CTA!1245)

### Performance Improvements

- [taped] Optimise DriveState queries (cta/CTA!1198)

### Other

- [Misc] Rename *.conf.example files to *.example.conf (cta/CTA!1279)

## 5.11.16.1-1

### Bug Fixes


stage/stage_updc.c
/*
 * $Id: stage_updc.c,v 1.2 2000/05/08 14:47:44 jdurand Exp $
 */

/*
 * Copyright (C) 1993-1999 by CERN/IT/PDP/DM
 * All rights reserved
 */

#ifndef lint
static char sccsid[] = "@(#)$RCSfile: stage_updc.c,v $ $Revision: 1.2 $ $Date: 2000/05/08 14:47:44 $ CERN IT-PDP/DM Jean-Damien Durand Jean-Philippe Baud";
#endif /* not lint */

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <fcntl.h>
#include <pwd.h>
#include <string.h>
#if defined(_WIN32)
#include <winsock2.h>
#else
#include <unistd.h>
#include <netinet/in.h>
#endif
#include "marshall.h"
#include "serrno.h"
#include "stage_api.h"
#include "stage.h"
#include "u64subr.h"

int copyrc_shift2castor _PROTO((int));

int copyrc_castor2shift(copyrc)
     int copyrc;
{
  /* Input  is a CASTOR return code */
  /* Output is a SHIFT  return code */
  switch (copyrc) {
  case ERTBLKSKPD:
    return(BLKSKPD);
  case ERTTPE_LSZ:
    return(TPE_LSZ);
  case ERTMNYPARY:
  case ETPARIT:
    return(MNYPARI);
  case ERTLIMBYSZ:
    return(LIMBYSZ);
  default:
    return(copyrc);
  }
}

int DLL_DECL stage_updc_filcp(stageid, copyrc, ifce, size, waiting_time, transfer_time, blksize, drive, fid, fseq, lrecl, recfm, path)
     char *stageid;
     int copyrc;
     char *ifce;
     u_signed64 size;
     int waiting_time;
     int transfer_time;
     int blksize;
     char *drive;
     char *fid;
     int fseq;
     int lrecl;
     char *recfm;
     char *path;
{
  int c;
  char *dp;
  int errflg = 0;
  gid_t gid;
  int key;
  int msglen;
  int nargs;
  int ntries = 0;
  char *p;
  struct passwd *pw;
  char *q, *q2;
  char *rbp;
  char repbuf[CA_MAXPATHLEN+1];
  int reqid;
  char *sbp;
  char sendbuf[REQBUFSZ];
  char *stghost;
  char tmpbuf[21];
  uid_t uid;
  char Zparm[CA_MAXSTGRIDLEN+1];

  uid = getuid();
  gid = getgid();
#if defined(_WIN32)
  if (uid < 0 || gid < 0) {
    serrno = SENOMAPFND;
    return (-1);
  }
#endif
  if (! stageid) {
    serrno = EFAULT;
    return (-1);
  }
  if (stageid[0] == '\0') {
    serrno = EFAULT;
    return (-1);
  }

  /* Init repbuf to null */
  repbuf[0] = '\0';

  /* check stager request id */

  if (strlen (stageid) <= CA_MAXSTGRIDLEN) {
    strcpy (Zparm, stageid);
    if ((p = strtok (Zparm, ".")) != NULL) {
      reqid = strtol (p, &dp, 10);
      if (*dp != '\0' || reqid <= 0 ||
          (p = strtok (NULL, "@")) == NULL) {
        errflg++;
      } else {
        key = strtol (p, &dp, 10);
        if (*dp != '\0' ||
            (stghost = strtok (NULL, " ")) == NULL) {
          errflg++;
        }
      }
    } else {
      errflg++;
    }
  } else
    errflg++;
  if (errflg) {
    serrno = EINVAL;
    return (-1);
  }

  /* Build request header */

  sbp = sendbuf;
  marshall_LONG (sbp, STGMAGIC);
  marshall_LONG (sbp, STAGEUPDC);
  q = sbp;	/* save pointer. The next field will be updated */
  msglen = 3 * LONGSIZE;
  marshall_LONG (sbp, msglen);

  /* Build request body */

  if ((pw = getpwuid (uid)) == NULL) {
    serrno = SENOMAPFND;
    return (-1);
  }
  marshall_STRING (sbp, pw->pw_name);	/* login name */
  marshall_WORD (sbp, uid);
  marshall_WORD (sbp, gid);
  q2 = sbp;	/* save pointer. The next field will be updated */
  nargs = 1;
  marshall_WORD (sbp, nargs);
  marshall_STRING (sbp, "stage_updc_filcp");

  marshall_STRING (sbp, "-Z");
  marshall_STRING (sbp, stageid);
  nargs += 2;
  if (blksize > 0) {
    sprintf (tmpbuf, "%d", blksize);
    marshall_STRING (sbp,"-b");
    marshall_STRING (sbp, tmpbuf);
    nargs += 2;
  }
  if (drive) {
    if (drive[0] != '\0') {
      marshall_STRING (sbp,"-D");
      marshall_STRING (sbp, drive);
      nargs += 2;
    }
  }
  if (recfm) {
    if (recfm[0] != '\0') {
      marshall_STRING (sbp,"-F");
      marshall_STRING (sbp, recfm);
      nargs += 2;
    }
  }
  if (fid) {
    if (fid[0] != '\0') {
      marshall_STRING (sbp,"-f");
      marshall_STRING (sbp, fid);
      nargs += 2;
    }
  }
  if (ifce) {
    if (ifce[0] != '\0') {
      marshall_STRING (sbp,"-I");
      marshall_STRING (sbp, ifce);
      nargs += 2;
    }
  }
  if (lrecl > 0) {
    sprintf (tmpbuf, "%d", lrecl);
    marshall_STRING (sbp,"-L");
    marshall_STRING (sbp, tmpbuf);
    nargs += 2;
  }
  if (size > 0) {
    u64tostr((u_signed64) size, tmpbuf, 0);
    marshall_STRING (sbp,"-s");
    marshall_STRING (sbp, tmpbuf);
    nargs += 2;
  }
  if (copyrc >= 0) {
    sprintf (tmpbuf, "%d", copyrc_castor2shift(copyrc));
    marshall_STRING (sbp, "-R");
    marshall_STRING (sbp, tmpbuf);
    nargs += 2;
  }
  if (transfer_time > 0) {
    sprintf (tmpbuf, "%d", transfer_time);
    marshall_STRING (sbp, "-T");
    marshall_STRING (sbp, tmpbuf);
    nargs += 2;
  }
  if (waiting_time > 0) {
    sprintf (tmpbuf, "%d", waiting_time);
    marshall_STRING (sbp, "-W");
    marshall_STRING (sbp, tmpbuf);
    nargs += 2;
  }
  if (fseq > 0) {
    sprintf (tmpbuf, "%d", fseq);
    marshall_STRING (sbp,"-q");
    marshall_STRING (sbp, tmpbuf);
    nargs += 2;
  }
  marshall_WORD (q2, nargs);	/* update nargs */

  msglen = sbp - sendbuf;
  marshall_LONG (q, msglen);	/* update length field */

  while (1) {
    c = send2stgd (stghost, sendbuf, msglen, 1, repbuf, sizeof(repbuf));
    if (c == 0) break;
    if (serrno != ESTNACT || ntries++ > MAXRETRY) break;
    sleep (RETRYI);
  }
  if (c == 0 && path != NULL && repbuf[0] != '\0') {
    rbp = repbuf;
    unmarshall_STRING (rbp, path);
  }
  return (c == 0 ? 0 : -1);
}

int DLL_DECL stage_updc_tppos(stageid, status, blksize, drive, fid, fseq, lrecl, recfm, path)
     char *stageid;
     int status;
     int blksize;
     char *drive;
     char *fid;
     int fseq;
     int lrecl;
     char *recfm;
     char *path;
{
  int c;
  char *dp;
  int errflg = 0;
  gid_t gid;
  int key;
  int msglen;
  int n;
  int nargs;
  int ntries = 0;
  char *p;
  struct passwd *pw;
  char *q, *q2;
  char *rbp;
  char repbuf[CA_MAXPATHLEN+1];
  int reqid;
  char *sbp;
  char sendbuf[REQBUFSZ];
  char *stghost;
  char tmpbuf[21];
  uid_t uid;
  char Zparm[CA_MAXSTGRIDLEN+1];

  uid = getuid();
  gid = getgid();
#if defined(_WIN32)
  if (uid < 0 || gid < 0) {
    serrno = SENOMAPFND;
    return (-1);
  }
#endif
  if (! stageid) {
    serrno = EFAULT;
    return (-1);
  }
  if (stageid[0] == '\0') {
    serrno = EFAULT;
    return (-1);
  }

  /* Init repbuf to null */
  repbuf[0] = '\0';

  /* check stager request id */

  if (strlen (stageid) <= CA_MAXSTGRIDLEN) {
    strcpy (Zparm, stageid);
    if ((p = strtok (Zparm, ".")) != NULL) {
      reqid = strtol (p, &dp, 10);
      if (*dp != '\0' || reqid <= 0 ||
          (p = strtok (NULL, "@")) == NULL) {
        errflg++;
      } else {
        key = strtol (p, &dp, 10);
        if (*dp != '\0' ||
            (stghost = strtok (NULL, " ")) == NULL) {
          errflg++;
        }
      }
    } else {
      errflg++;
    }
  } else
    errflg++;
  if (errflg) {
    serrno = EINVAL;
    return (-1);
  }

  /* Build request header */

  sbp = sendbuf;
  marshall_LONG (sbp, STGMAGIC);
  marshall_LONG (sbp, STAGEUPDC);
  q = sbp;	/* save pointer. The next field will be updated */
  msglen = 3 * LONGSIZE;
  marshall_LONG (sbp, msglen);

  /* Build request body */

  if ((pw = getpwuid (uid)) == NULL) {
    serrno = SENOMAPFND;
    return (-1);
  }
  marshall_STRING (sbp, pw->pw_name);	/* login name */
  marshall_WORD (sbp, uid);
  marshall_WORD (sbp, gid);
  q2 = sbp;	/* save pointer. The next field will be updated */
  nargs = 1;
  marshall_WORD (sbp, nargs);
  marshall_STRING (sbp, "stage_updc_tppos");

  marshall_STRING (sbp, "-Z");
  marshall_STRING (sbp, stageid);
  nargs += 2;
  if (blksize > 0) {
    sprintf (tmpbuf, "%d", blksize);
    marshall_STRING (sbp,"-b");
    marshall_STRING (sbp, tmpbuf);
    nargs += 2;
  }
  if (drive) {
    if (drive[0] != '\0') {
      marshall_STRING (sbp,"-D");
      marshall_STRING (sbp, drive);
      nargs += 2;
    }
  }
  if (recfm) {
    if (recfm[0] != '\0') {
      marshall_STRING (sbp,"-F");
      marshall_STRING (sbp, recfm);
      nargs += 2;
    }
  }
  if (fid) {
    if (fid[0] != '\0') {
      marshall_STRING (sbp,"-f");
      marshall_STRING (sbp, fid);
      nargs += 2;
    }
  }
  if (lrecl > 0) {
    sprintf (tmpbuf, "%d", lrecl);
    marshall_STRING (sbp,"-L");
    marshall_STRING (sbp, tmpbuf);
    nargs += 2;
  }
  if (fseq > 0) {
    sprintf (tmpbuf, "%d", fseq);
    marshall_STRING (sbp,"-q");
    marshall_STRING (sbp, tmpbuf);
    nargs += 2;
  }
  if (status == ETFSQ) {
    n = 211;	/* emulate old (SHIFT) ETFSQ */
    sprintf (tmpbuf, "%d", n);
    marshall_STRING (sbp, "-R");
    marshall_STRING (sbp, tmpbuf);
    nargs += 2;
  }
  marshall_WORD (q2, nargs);	/* update nargs */

  msglen = sbp - sendbuf;
  marshall_LONG (q, msglen);	/* update length field */

  while (1) {
    c = send2stgd (stghost, sendbuf, msglen, 1, repbuf, sizeof(repbuf));
    if (c == 0) break;
    if (serrno != ESTNACT || ntries++ > MAXRETRY) break;
    sleep (RETRYI);
  }
  if (c == 0 && path != NULL && repbuf[0] != '\0') {
    rbp = repbuf;
    unmarshall_STRING (rbp, path);
  }
  return (c == 0 ? 0 : -1);
}

int DLL_DECL stage_updc_user(stghost,linkfile)
     char *stghost;
     char *linkfile;
{
  int c;
  gid_t gid;
  int msglen;
  int nargs;
  int ntries = 0;
  struct passwd *pw;
  char *q, *q2;
  char repbuf[CA_MAXPATHLEN+1];
  char *sbp;
  char sendbuf[REQBUFSZ];
  uid_t uid;
  
  uid = getuid();
  gid = getgid();
#if defined(_WIN32)
  if (uid < 0 || gid < 0) {
    serrno = SENOMAPFND;
    return (-1);
  }
#endif
  if (! linkfile) {
    serrno = EFAULT;
    return (-1);
  }
  if (linkfile[0] == '\0') {
    serrno = EFAULT;
    return (-1);
  }

  /* Init repbuf to null */
  repbuf[0] = '\0';

  /* Build request header */

  sbp = sendbuf;
  marshall_LONG (sbp, STGMAGIC);
  marshall_LONG (sbp, STAGEUPDC);
  q = sbp;	/* save pointer. The next field will be updated */
  msglen = 3 * LONGSIZE;
  marshall_LONG (sbp, msglen);

  /* Build request body */

  if ((pw = getpwuid (uid)) == NULL) {
    serrno = SENOMAPFND;
    return (-1);
  }
  marshall_STRING (sbp, pw->pw_name);	/* login name */
  marshall_WORD (sbp, uid);
  marshall_WORD (sbp, gid);
  q2 = sbp;	/* save pointer. The next field will be updated */
  nargs = 1;
  marshall_WORD (sbp, nargs);
  marshall_STRING (sbp, "stage_updc_user");

  marshall_STRING (sbp, linkfile);
  nargs += 1;
  marshall_WORD (q2, nargs);	/* update nargs */

  msglen = sbp - sendbuf;
  marshall_LONG (q, msglen);	/* update length field */

  while (1) {
    c = send2stgd (stghost, sendbuf, msglen, 1, repbuf, sizeof(repbuf));
    if (c == 0) break;
    if (serrno != ESTNACT || ntries++ > MAXRETRY) break;
    sleep (RETRYI);
  }
  return (c == 0 ? 0 : -1);
}
